python/lsst/pipe/base/_task_metadata.py

```diff
@@ -0,0 +1,427 @@
+# This file is part of pipe_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["TaskMetadata"]
+
+import numbers
+import itertools
+import warnings
+from collections.abc import Sequence
+from deprecated.sphinx import deprecated
+
+from typing import Dict, List, Union, Any, Mapping
+from pydantic import BaseModel, StrictInt, StrictFloat, StrictBool, StrictStr, Field
+
+_DEPRECATION_REASON = "Will be removed after v25."
+_DEPRECATION_VERSION = "v24"
+
+# The types allowed in a Task metadata field are restricted
+# to allow predictable serialization.
+_ALLOWED_PRIMITIVE_TYPES = (str, float, int, bool)
+
+
+def _isListLike(v):
+    return isinstance(v, Sequence) and not isinstance(v, str)
+
+
+class TaskMetadata(BaseModel):
+    """Dict-like object for storing task metadata.
+    Metadata can be stored at two levels: single task or task plus subtasks.
+    The later is called full metadata of a task and has a form
+
+        topLevelTaskName:subtaskName:subsubtaskName.itemName
+
+    Metadata item key of a task (`itemName` above) must not contain `.`,
```

Comment:
Hasn't this changed?

Reply:
Right. Before this most recent change only a single `.` was allowed.